File: exam/src/services/validator/DocumentValidatorFactory.ts
import { Document } from "../../models/Document";
import { ContractValidator } from "../validator/ContractValidator";
import { FinancialReportValidator } from "../validator/FinancialReportValidator";
import { ProposalValidator } from "../validator/ProposalValidator";
import { DocumentValidator } from "../validator/DocumentValidator";

export class DocumentValidatorFactory {
    private static readonly DOCUMENT_TYPE_CONTRACT = "Contrato Legal";
    private static readonly DOCUMENT_TYPE_FINANCIAL_REPORT = "Reporte Financiero";
    private static readonly DOCUMENT_TYPE_PROPOSAL = "Propuesta";

    static validate(document: Document): DocumentValidator {
        switch (document.getDocumentType()) {
            case this.DOCUMENT_TYPE_CONTRACT:
                return new ContractValidator().validate(document);
            case this.DOCUMENT_TYPE_FINANCIAL_REPORT:
                return new FinancialReportValidator().validate(document);
            case this.DOCUMENT_TYPE_PROPOSAL:
                return new ProposalValidator().validate(document);
            default:
                throw new Error("This file type is not supported");
        }
    }
}

File: exam/src/validator/DocumentValidatorFactory.java
package validator;

import enums.DocumentType;
import validator.documentValidators.*;

public class DocumentValidatorFactory {

    public static DocumentValidator getDocumentValidator(DocumentType type) {

        return switch (type) {
            case CONTRACT -> new ContractValidator();
            case FINANCIAL_REPORT -> new FinancialReportValidator();
            case PROPOSAL -> new ProposalValidator();
        };
    }
}
